### PR TITLE
[storage/index] simplify+optimize Cursor

### PR DESCRIPTION
**Pull request by roberto-bayardo**
## Summary

Updates `storage::index` cursor mutation to operate in place, avoiding the previous detach/reattach flow while preserving cursor semantics. Inserted collisions are now (typically) stored and returned in newest-first order, though contract continues to provide no ordering guarantee among colliding keys.

## Changes

- Reworked shared cursor state around `prev` / `current` pointers.
- Reduced unsafe surface to one centralized mutable pointer dereference, with pointer creation restricted to `&mut Record`.
- Fixed `insert_and_prune` so vacant entries are not created when the incoming value is already prunable.
- Made `remove` drain owned collision chains directly while keeping metrics accurate.
- Updated ordered, unordered, and partitioned index tests for newest-first value order.
- Added regression coverage for prunable vacant inserts and cursor delete/insert edge cases.
- Added `insert_and_prune` benchmark

## Results

  | Benchmark | Before | After | Change |                                                                                                                                        
  |-----------|--------|-------|--------|                   
  | `index::insert_and_prune/items=10000` | 191.5 µs | 77.8 µs | **-59.3%** |
  | `index::insert_and_prune/items=50000` | 1.163 ms | 479.1 µs | **-58.8%** |      

 | Benchmark | Time | Change vs main |
  |-----------|------|----------------|
  | `qmdb::apply_batch/direct` | 746 µs | **-14.0%** |
  | `qmdb::apply_batch/uncomm_ancestor` | 2.39 ms | **-5.3%** |
  | `qmdb::apply_batch/comm_ancestor` | 1.49 ms | **-6.5%** |
  | `qmdb::apply_batch/comm_uncomm_chain` | 2.94 ms | **-15.7%** |                                                                                                                     
  | `qmdb::apply_batch/multi_uncomm` | 3.71 ms | **-9.1%** |

### CODE REVIEW

**Comment by roberto-bayardo** [comment]
bugbot run

File: storage/src/index/storage.rs [hunk 1]
```diff
@@ -286,69 +226,18 @@ impl<V: Eq + Send + Sync, E: IndexEntry<V>> CursorTrait for Cursor<'_, V, E> {
...
-    V: Eq + Send + Sync,
-    E: IndexEntry<V>,
-{
-}
+// SAFETY: `prev` and `current` are only bookkeeping pointers into the linked list owned by
+// `entry`. Moving or sharing the cursor does not grant access to those records without `&mut self`,
+// and `entry` keeps the list alive and exclusively borrowed for the cursor's lifetime.
+unsafe impl<V: Eq + Send + Sync, E: IndexEntry<V>> Send for Cursor<'_, V, E> {}
+// SAFETY: See `Send` safety documentation.
+unsafe impl<V: Eq + Send + Sync, E: IndexEntry<V>> Sync for Cursor<'_, V, E> {}
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/index/storage.rs]: no removing these results in compile errors

File: storage/src/index/storage.rs [hunk 2]
```diff
@@ -27,256 +27,196 @@ pub(super) trait IndexEntry<V: Eq + Send + Sync>: Send + Sync {
...
+/// Tracks `prev` (for relinking on delete) and `current` (last item returned by `next`).
+/// The next element to visit is derived from `current.next` (or the entry head when
+/// `current` is `None`), so no separate `upcoming` pointer is needed.
+///
+/// Invariants:
+/// - `entry` owns the linked list and keeps it exclusively borrowed for the cursor's lifetime.
+/// - `prev` and `current`, when present, point into that list.
+/// - `prev` and `current` are created only from exclusive references through `record_ptr`.
+/// - When both are present, `prev.next` owns `current`.
+/// - After deleting a node, `current` is moved back to the previous live node or cleared.
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/index/storage.rs]: fixed

**Comment by patrick-ogrady** [comment]
Found an issue:

```
  This specific retained-suffix ordering bug came from this PR’s switch to insert_and_prune. Main
  already had related gap-fill logic, but this PR made older gap entries append after newer retained
  entries. The fix replays the gap in reverse with front insertion, preserving log order for both
  “all repeated writes in gap” and “older write in gap, newer write retained” cases.

+    /// Regression: after restart, the snapshot can contain the newer write for
+    /// a repeated key. If rewind restores an older write for that same key, the
+    /// older write must be checked first, matching the pre-restart snapshot.
```

Patch:

```
diff --git a/storage/src/qmdb/immutable/mod.rs b/storage/src/qmdb/immutable/mod.rs
index 165ff02f8..2e789c5d7 100644
--- a/storage/src/qmdb/immutable/mod.rs
+++ b/storage/src/qmdb/immutable/mod.rs
@@ -589,10 +589,12 @@ where
         if rewind_floor < old_floor {
             let reader = self.journal.journal.reader().await;
             let gap_end = core::cmp::min(*old_floor, rewind_size);
-            for loc in *rewind_floor..gap_end {
+            for loc in (*rewind_floor..gap_end).rev() {
                 if let Operation::Set(key, _) = reader.read(loc).await? {
-                    self.snapshot
-                        .insert_and_prune(&key, Location::new(loc), |_| false);
+                    let loc = Location::new(loc);
+                    if !self.snapshot.get(&key).any(|existing| *existing == loc) {
+                        self.snapshot.insert(&key, loc);
+                    }
                 }
             }
         }
@@ -2928,9 +2930,9 @@ pub(super) mod test {
         db.destroy().await.unwrap();
     }
 
-    /// Regression: rewind-after-reopen with a repeated key in the floor gap.
-    /// The gap-fill must maintain the same ordering as the live `apply_batch`
-    /// path so `get()` returns the same value.
+    /// Regression: rewind-after-reopen with a repeated key split across the
+    /// floor gap and retained suffix. The gap-fill must preserve log order
+    /// relative to entries that survived the higher-floor reopen.
     pub(crate) async fn test_immutable_rewind_after_reopen_repeated_key_gap<F: Family, V, C>(
         context: deterministic::Context,
         open_db: impl Fn(
@@ -2946,29 +2948,37 @@ pub(super) mod test {
         let key = Sha256::fill(7u8);
         let v1 = Sha256::fill(17u8);
         let v2 = Sha256::fill(18u8);
-        let k3 = Sha256::fill(8u8);
-        let v3 = Sha256::fill(19u8);
+        let marker = Sha256::fill(8u8);
+        let marker_value = Sha256::fill(19u8);
+        let k3 = Sha256::fill(9u8);
+        let v3 = Sha256::fill(20u8);
 
         // Commit A: Set(key, v1) with floor=0.
         commit_sets(&mut db, [(key, v1)], None).await;
 
-        // Commit B: Set(key, v2) with floor=0. get() returns v1 (earliest).
-        commit_sets(&mut db, [(key, v2)], None).await;
-        let second_size = db.bounds().await.end;
+        // Commit B: advances the target floor to key's first write.
+        commit_sets_with_floor(&mut db, [(marker, marker_value)], None, Location::new(1)).await;
+
+        // Commit C: Set(key, v2) at a location that will remain in the
+        // reopened snapshot. The rewind target still includes the older write.
+        commit_sets_with_floor(&mut db, [(key, v2)], None, Location::new(1)).await;
+        let target_size = db.bounds().await.end;
         assert_eq!(db.get(&key).await.unwrap(), Some(v1));
 
-        // Commit C: raises floor above both earlier writes.
-        commit_sets_with_floor(&mut db, [(k3, v3)], None, second_size).await;
+        // Commit D: raises the latest floor to key's second write, so reopen
+        // retains v2 but excludes v1.
+        commit_sets_with_floor(&mut db, [(k3, v3)], None, Location::new(5)).await;
         db.sync().await.unwrap();
 
-        // Reopen: snapshot rebuilt from floor=second_size, key excluded.
+        // Reopen: snapshot rebuilt from floor=5, so key resolves to v2.
         drop(db);
         let mut db = open_db(context.child("second")).await;
-        assert!(db.get(&key).await.unwrap().is_none());
+        assert_eq!(db.get(&key).await.unwrap(), Some(v2));
         assert_eq!(db.get(&k3).await.unwrap(), Some(v3));
 
-        // Rewind to commit B: gap fill re-inserts both Set(key,...) entries.
-        db.rewind(second_size).await.unwrap();
+        // Rewind to commit C: v2 is retained and v1 is restored from the gap.
+        // Reads must match the live target snapshot, where the earlier write wins.
+        db.rewind(target_size).await.unwrap();
         assert_eq!(db.get(&key).await.unwrap(), Some(v1));
 
         db.destroy().await.unwrap();
```

**Comment by roberto-bayardo** [comment]
Yeah this is the issue the QED bot turned up, looks like my fix wasn't a full fix though. (Admittedly I half hearted it since there really aren't supposed to be duplicate keys in the first place in an "immutable" db.). Any fix relying on insert() is going to be error prone since technically it doesn't guarantee the resulting order among collisions, but this patch is good enough for now (modulo one line noted below).

> Found an issue:
> 
> ```
>   This specific retained-suffix ordering bug came from this PR’s switch to insert_and_prune. Main
>   already had related gap-fill logic, but this PR made older gap entries append after newer retained
>   entries. The fix replays the gap in reverse with front insertion, preserving log order for both
>   “all repeated writes in gap” and “older write in gap, newer write retained” cases.
> 
> +    /// Regression: after restart, the snapshot can contain the newer write for
> +    /// a repeated key. If rewind restores an older write for that same key, the
> +    /// older write must be checked first, matching the pre-restart snapshot.
> ```
> 
> Patch:
> 
> ```
> diff --git a/storage/src/qmdb/immutable/mod.rs b/storage/src/qmdb/immutable/mod.rs
> index 165ff02f8..2e789c5d7 100644
> --- a/storage/src/qmdb/immutable/mod.rs
> +++ b/storage/src/qmdb/immutable/mod.rs
> @@ -589,10 +589,12 @@ where
>          if rewind_floor < old_floor {
>              let reader = self.journal.journal.reader().await;
>              let gap_end = core::cmp::min(*old_floor, rewind_size);
> -            for loc in *rewind_floor..gap_end {
> +            for loc in (*rewind_floor..gap_end).rev() {
>                  if let Operation::Set(key, _) = reader.read(loc).await? {
> -                    self.snapshot
> -                        .insert_and_prune(&key, Location::new(loc), |_| false);
> +                    let loc = Location::new(loc);
> +                    if !self.snapshot.get(&key).any(|existing| *existing == loc) {
>```

^^ This line isn't needed (duplicate locations aren't allowed). but otherwise I've applied this patch, and added a distinct regression test rather than patch the original one.

>```
> +                        self.snapshot.insert(&key, loc);
> +                    }
>```

**Review by danlaine — DISMISSED** [review]

File: storage/src/index/benches/insert_and_prune.rs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+) -> Duration {
+    // Seed the index with initial values.
+    for (k, v) in kvs {
+        index.insert(k, *v);
+    }
+
+    // Overwrite every key using insert_and_prune: prune the old value, insert the new one.
+    let start = Instant::now();
+    for (k, v) in kvs {
+        index.insert_and_retain(k, *v + 1, |old| *old == *v);
```
[the remark this answers is not in the document]
Comment (danlaine) [reply storage/src/index/benches/insert_and_prune.rs]: fixed in [4686962](https://github.com/commonwarexyz/monorepo/pull/3760/commits/4686962393b281160df628b310906432a62f8c65)

**Review by danlaine — DISMISSED** [review]

**Review by danlaine — APPROVED** [review]
